### ISSUE TITLE
`unused_patterns` lint 

### ISSUE THREAD

**Issue by nikomatsakis** [rust-lang/rust#54538]
As @scottmcm [pointed out here](https://github.com/rust-lang/rust/issues/51087#issuecomment-397918200), the `unused_parens` lints should be extended to also catch unused parentheses in patterns:

```rust
    match (1) { // Warns, as expected
        (_) => {} // Doesn't warn, but should for consistency
    }
```

[Play](https://play.rust-lang.org/?gist=9f08276033eab518cd8055ff148c266a&version=nightly)

**Comment by nikomatsakis** [comment]
Here are some tips for how to update that lint. These aren't full mentoring notes but they might get you started. 

The lint itself is declared here:

https://github.com/rust-lang/rust/blob/3bc2ca7e4f8507f82a4c357ee19300166bcd8099/src/librustc_lint/unused.rs#L258-L262

The meat of it is found in the `check_expr` callback here, which is invoked for every expression:

https://github.com/rust-lang/rust/blob/3bc2ca7e4f8507f82a4c357ee19300166bcd8099/src/librustc_lint/unused.rs#L326

As you can see, it checks for various places where a lint would not be needed, such as in the body if an `If`, and then invokes this routine [`check_unused_patterns_core`](https://github.com/rust-lang/rust/blob/3bc2ca7e4f8507f82a4c357ee19300166bcd8099/src/librustc_lint/unused.rs#L326) to issue a warning.

For patterns, I imagine we'll want to do something similar, but adding a `check_pat` method ([this is the definition from the trait](https://doc.rust-lang.org/nightly/nightly-rustc/rustc/lint/trait.EarlyLintPass.html#method.check_pat)). 

That said, I think this is quite a bit harder than the stabilization itself. If the person who handles this issue wants to do it, seems great, but if they are not familiar with the compiler, I think we could leave it out from the first PR, and then move this to a distinct issue to be added at some later time.

**Comment by kleimkuhler** [comment]
I can pick this up. The mentoring notes provide enough for me to get started. I will reach out if I need some help or clarification!

**Comment by nikomatsakis** [comment]
@kleimkuhler Great! If you have any further problems, please don't hestitate to ping on [Zulip](https://rust-lang.zulipchat.com/) or [Discord](https://discord.gg/rust-lang).

**Comment by nikomatsakis** [comment]
[I started a thread on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/.2354538.20.20.60unused_patterns.60.20lint), for future reference.
